File: Training/GameManagerScript/Neuron.py
import random
import math


class Neuron:
    link_break_chance = (1, 1000)
    link_create_chance = (4, 1000)

    def __init__(self, i=0):
        self.links = []
        self.id = i
        self.value = random.random() * 2 - 1

    def load_from_line(self, line):
        self.links = []
        elems = [float(i) for i in line.split(",")]
        self.id = int(elems[0])
        self.value = int(elems[1])
        for i in range(int(elems[2])):
            self.links.append([elems[3 + i * 2], elems[4 + i * 2]])

    def resolve_links(self, elems):
        for k, link in enumerate(self.links):
            if isinstance(link[1], int):
                link[1], = (i for i in elems if i.id == link[1])

    def mutate(self, others):
        to_remove = []
        for i, link in enumerate(self.links):
            # Mutate links
            if random.randint(0, 1) == 0:
                link[0] += math.exp(random.random() / -2) / 4
            else:
                link[0] -= math.exp(random.random() / -2) / 4
            link[0] = max(min(1, link[0]), -1)
            if len(self.links) != 0 and random.randrange(0, self.link_break_chance[1]) < self.link_break_chance[0]:
                to_remove.insert(0, i)
        if random.randint(0, 1) == 0:
            self.value += math.exp(random.random() / -2) / 4
        else:
            self.value -= math.exp(random.random() / -2) / 4
        for i in to_remove:
            self.links.pop(i)
        # Add some links
        while random.randrange(0, self.link_create_chance[1]) < self.link_create_chance[0]:
            self.links.append([random.random() * 2 - 1, random.choice(tuple(filter(lambda x: self.id not in x.get_dependencies(), others)))])

    def get_value(self):
        if len(self.links) == 0:
            return self.value
        val = sum(link[0] * link[1].get_value() for link in self.links) + self.value
        return min(max(val, -1), 1)

    def get_dependencies(self):
        return set(j for i in ((link[1].id, *link[1].get_dependencies()) for link in self.links) for j in i)

    def copy(self):
        me = Neuron(self.id)
        me.value = self.value
        me.links = [[link[0], link[1].id] for link in self.links]
        return me

    def save(self, stream):
        stream.write(",".join(
            (str(self.id), str(len(self.links)), str(self.value), *(",".join((str(link[0]), str(link[1].id))) for link in self.links))
        ) + "\n")

File: Training/AITrainerScriptCPP/Neuron.cpp
//
// Created by PinkySmile on 22/05/2021.
//

#include <cmath>
#include <sstream>
#include <fstream>
#include <iostream>
#include <algorithm>
#ifdef _WIN32
#include <Windows.h>
#else
using std::max;
using std::min;
#endif
#include "Neuron.hpp"

#define LINK_BREAK_CHANCE_NUM 1
#define LINK_BREAK_CHANCE_DEN 1000
#define LINK_CREATE_CHANCE_NUM 4
#define LINK_CREATE_CHANCE_DEN 1000

namespace Trainer
{
	Neuron::Neuron(unsigned int i) :
		_id(i),
		_value(2.f * rand() / RAND_MAX - 1)
	{
	}

	Neuron::Neuron(const Neuron &other)
	{
		this->_id = other._id;
		this->_value = other._value;
		this->_links.reserve(other._links.size());
		for (auto &link : other._links)
			this->_links.push_back({
				link.id,
				link.weight,
				nullptr
			});
	}

	void Neuron::loadFromLine(const std::string &line)
	{
		unsigned i;
		std::string token;
		std::stringstream stream{line};

		std::getline(stream, token, ',');
		this->_id = std::stoul(token);

		std::getline(stream, token, ',');
		i = std::stoul(token);

		std::getline(stream, token, ',');
		this->_value = std::stof(token);

		this->_links.clear();
		this->_links.reserve(i);
		while (i--) {
			std::string weight;
			std::string id;

			std::getline(stream, weight, ',');
			std::getline(stream, id, ',');
			this->_links.push_back({
				(unsigned)std::stoul(id),
				std::stof(weight),
				nullptr
			});
		}
	}

	void Neuron::resolveLinks(const std::vector<std::unique_ptr<Neuron>> &elems)
	{
		for (int i = 0; i < this->_links.size(); i++) {
			auto &link = this->_links[i];

			if (!link.neuron)
				for (auto &neuron : elems)
					if (neuron->getId() == link.id) {
						link.neuron = &*neuron;
						break;
					}
			if (!link.neuron) {
				this->_links.erase(this->_links.begin() + i);
				i--;
			}
		}
	}

	void Neuron::mutate(const std::vector<std::unique_ptr<Neuron>> &others)
	{
		for (int i = 0; i < this->_links.size(); i++) {
			auto &link = this->_links[i];

			// Mutate links
			if ((rand() & 1) == 0)
				link.weight += std::exp(rand() / -2.f / RAND_MAX) / 4;
			else
				link.weight -= std::exp(rand() / -2.f / RAND_MAX) / 4;
			link.weight = max(min(1.f, link.weight), -1.f);
			if (!this->_links.empty() && rand() % LINK_BREAK_CHANCE_DEN < LINK_BREAK_CHANCE_NUM) {
				this->_links.erase(this->_links.begin() + i);
				i--;
			}
		}
		if ((rand() & 1) == 0)
			this->_value += std::exp(rand() / -2.f / RAND_MAX) / 4;
		else
			this->_value -= std::exp(rand() / -2.f / RAND_MAX) / 4;
		// Add some links
		while (rand() % LINK_CREATE_CHANCE_DEN < LINK_CREATE_CHANCE_NUM) {
			std::vector<Neuron *> result;

			for (auto &neur : others) {
				auto deps = neur->getDependencies();

				if (std::find(deps.begin(), deps.end(), this->_id) == deps.end())
					result.push_back(&*neur);
			}
			if (!result.empty()) {
				auto r = result[rand() % result.size()];

				this->_links.push_back({
					r->getId(),
					rand() * 2.f / RAND_MAX - 1,
					r
				});
			} else
				break;
		}
	}

	float Neuron::getValue()
	{
		float result = 0;

		for (auto &link : this->_links)
			result += link.weight * link.neuron->getValue();
		return min(max(result + this->_value, -1.f), 1.f);
	}

	std::set<unsigned> Neuron::getDependencies()
	{
		std::set<unsigned> deps;

		for (auto &link : this->_links) {
			const auto &innerDeps = link.neuron->getDependencies();

			deps.insert(innerDeps.begin(), innerDeps.end());
		}
		deps.insert(this->_id);
		return deps;
	}

	void Neuron::save(std::ofstream &stream)
	{
		stream << this->_id << ',';
		stream << this->_links.size() << ',';
		stream << this->_value;
		for (auto &link : this->_links)
			stream << ',' << link.weight << ',' << link.id;
		stream << std::endl;
	}

	unsigned int Neuron::getId() const
	{
		return this->_id;
	}

	void Neuron::setValue(float value)
	{
		this->_value = value;
	}

	void Neuron::link(Neuron *neuron, float weight)
	{
		this->_links.push_back({
			neuron->getId(),
			weight,
			neuron
		});
	}
}